File: haskell/glue/src/Glue/Lib/Builtin/Concat.hs
module Glue.Lib.Builtin.Concat where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

concat :: IR Eval
concat = NativeFunc concatImpl

concatImpl :: IR Eval -> Eval (IR Eval)
concatImpl ir1 = pure $ NativeFunc (concatTo ir1)

concatTo :: IR Eval -> IR Eval -> Eval (IR Eval)
concatTo ir1 ir2 = case (ir1, ir2) of
    (Object xs, Object ys) -> pure $ Object (xs <> ys)
    (String xs, String ys) -> pure $ String (xs <> ys)
    (List xs, List ys) -> pure $ List (xs <> ys)
    _ -> throwError $ wrongArgumentType ["Lists, objects or strings required"]

File: dart/glue/lib/src/lib/builtin/concat.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Concat function - concatenates two lists, objects or strings
/// Mirrors Haskell Glue.Lib.List.Concat.concat exactly
Ir concat = IrNativeFunc(concatImpl);

/// Concat function implementation
/// Mirrors Haskell Glue.Lib.List.Concat.concatImpl exactly
Eval<Ir> concatImpl(Ir ir1) {
  return Eval.pure(IrNativeFunc(concatWith(ir1)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Builtin.Concat.concatWith exactly
Eval<Ir> Function(Ir) concatWith(Ir ir1) {
  return (Ir ir2) => switch ((ir1, ir2)) {
    (IrObject(properties: final p1), IrObject(properties: final p2)) =>
      Eval.pure(IrObject(p1.addAll(p2, keepOrder: true).unlock)),
    (IrString(value: final v1), IrString(value: final v2)) => Eval.pure(
      IrString(v1 + v2),
    ),
    (IrList(elements: final e1), IrList(elements: final e2)) => Eval.pure(
      IrList([...e1, ...e2]),
    ),
    _ => throwError(wrongArgumentType(['Lists, objects or strings required'])),
  };
}
